src/librbd/api/Group.cc

```diff
@@ -1226,24 +1216,118 @@ int Group<I>::snap_rename(librados::IoCtx& group_ioctx, const char *group_name,
   return 0;
 }
 
+int convert_cls_rbd_GroupSnapshot_to_librbd_group_snap_info2_t(
+    librados::IoCtx& group_ioctx, const std::string& group_id,
+    const cls::rbd::GroupSnapshot* cls_group_snap,
+    group_snap_info2_t* group_snap)
+{
+  std::vector<group_image_snap_info_t> image_snaps;
+  image_snaps.reserve(cls_group_snap->snaps.size());
+
+  for (const auto& snap : cls_group_snap->snaps) {
+    librbd::IoCtx image_ioctx;
+    int r = util::create_ioctx(group_ioctx, "image", snap.pool, {},
+                               &image_ioctx);
+    if (r < 0) {
+      return r;
+    }
+
+    std::string image_name;
+    r = cls_client::dir_get_name(&image_ioctx, RBD_DIRECTORY, snap.image_id,
+                                 &image_name);
+    if (r < 0) {
+      return r;
+    }
+
+    image_snaps.emplace_back(image_name, snap.pool, snap.snap_id);
```

Comment:
This doesn't build, at least with clang:
```
[2454/2755] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o
FAILED: src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/fmt/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/include -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++20 -MD -MT src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o -MF src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o.d -o src/librbd/CMakeFiles/rbd_internal.dir/api/Group.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc:4:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Cond.h:19:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Clock.h:18:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/utime.h:28:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/types.h:21:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/uuid.h:9:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/include/encoding.h:18:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/set:60:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tree.h:67:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/ext/alloc_traits.h:34:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/alloc_traits.h:518:4: error: no matching function for call to 'construct_at'
          std::construct_at(__p, std::forward<_Args>(__args)...);
          ^~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/vector.tcc:117:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<librbd::group_image_snap_info_t>>::construct<librbd::group_image_snap_info_t, std::basic_string<char> &, const long &, const snapid_t &>' requested here
            _Alloc_traits::construct(this->_M_impl, this->_M_impl._M_finish,
                           ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/api/Group.cc:1242:17: note: in instantiation of function template specialization 'std::vector<librbd::group_image_snap_info_t>::emplace_back<std::basic_string<char> &, const long &, const snapid_t &>' requested here
    image_snaps.emplace_back(image_name, snap.pool, snap.snap_id);
                ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_construct.h:94:5: note: candidate template ignored: substitution failure [with _Tp = librbd::group_image_snap_info_t, _Args = <std::basic_string<char> &, const long &, const snapid_t &>]: no matching constructor for initialization of 'librbd::group_image_snap_info_t'
    construct_at(_Tp* __location, _Args&&... __args)
    ^
1 error generated.
```
I think the issue is that `ImageSnapshotSpec`'s `snap_id` and `group_image_snap_info_t`'s `snap_id` aren't the same type (`snapid_t` which is a struct vs `uint64_t`) and the implicit conversion may not be picked up in this context.

Reply:
Let's see if explicit conversion to uint64_t fixes it

Reply:
Still fails

Reply:
Sigh, Clang 14 just doesn't support this C++20 feature.  The following trivial example doesn't compile for me either:
```
#include <vector>

struct foo {
    int i;
};

int main() {
    std::vector<foo> v;
    v.emplace_back(1);
}
```
According to https://en.cppreference.com/w/cpp/compiler_support/20, "Parenthesized initialization of aggregates" is supported only starting with Clang 16.

Reply:
Switched back to using `push_back()`